Source/Initialization/ExternalField.cpp

```diff
@@ -0,0 +1,179 @@
+/* Copyright 2023 Luca Fedeli
+ *
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ExternalField.H"
+
+#include "Utils/TextMsg.H"
+#include "Utils/Parser/ParserUtils.H"
+
+#include <ablastr/warn_manager/WarnManager.H>
+
+#include <algorithm>
+#include <vector>
+
+namespace
+{
+
+    enum class EMFieldType{E, B};
+
+    template <EMFieldType T>
+    ExternalFieldType string_to_external_field_type(std::string s)
+    {
+        std::transform(s.begin(), s.end(), s.begin(), ::tolower);
+
+        if ( s.empty() || s == "default")
+            return ExternalFieldType::default_zero;
+        else if ( s == "constant")
+            return ExternalFieldType::constant;
+        else if ( s == "parse_b_ext_grid_function" || s == "parse_e_ext_grid_function")
+            return ExternalFieldType::parse_ext_grid_function;
+        else if ( s == "read_from_file")
+            return ExternalFieldType::read_from_file;
+        else
+            WARPX_ABORT_WITH_MESSAGE(
+                "'" + s + "' is an unknown external field type!");
+
+        if constexpr (T == EMFieldType::E){
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(s != "parse_b_ext_grid_function",
+                "parse_B_ext_grid_function can be used only for B_ext_grid_init_style");
+        }
+        else{
+            WARPX_ALWAYS_ASSERT_WITH_MESSAGE(s != "parse_e_ext_grid_function",
+                "parse_E_ext_grid_function can be used only for E_ext_grid_init_style");
+        }
+
+        return ExternalFieldType::default_zero;
+    }
+}
+
+std::unique_ptr<ExternalFieldParams> ReadExternalFieldParams(const amrex::ParmParse& pp_warpx)
```

Comment:
Instead of a separate routine, can this be the constructor of `ExternalFieldParams`?

Reply:
yes. That's a good idea.